rpc/src/rpc_subscriptions.rs

```diff
@@ -430,40 +418,36 @@ fn filter_program_results(
             .iter()
             .all(|filter_type| filter_type.allows(account))
     });
-    let accounts: Box<dyn Iterator<Item = RpcKeyedAccount>> =
-        if is_known_spl_token_id(&params.pubkey)
-            && params.encoding == UiAccountEncoding::JsonParsed
-            && !accounts_is_empty
-        {
-            Box::new(get_parsed_token_accounts(bank, keyed_accounts))
-        } else {
-            Box::new(
-                keyed_accounts.map(move |(pubkey, account)| RpcKeyedAccount {
-                    pubkey: pubkey.to_string(),
-                    account: UiAccount::encode(&pubkey, &account, encoding, None, None),
-                }),
-            )
-        };
-    (accounts, last_notified_slot)
+    let (accounts, token_accounts) = if is_known_spl_token_id(&params.pubkey)
+        && params.encoding == UiAccountEncoding::JsonParsed
+        && !accounts_is_empty
+    {
+        let token_accounts = get_parsed_token_accounts(bank, keyed_accounts);
+        (None, Some(token_accounts))
+    } else {
+        let accounts = keyed_accounts.map(move |(pubkey, account)| RpcKeyedAccount {
+            pubkey: pubkey.to_string(),
+            account: UiAccount::encode(&pubkey, &account, encoding, None, None),
+        });
+        (Some(accounts), None)
+    };
+    let accounts = accounts.into_iter().flatten();
+    let token_accounts = token_accounts.into_iter().flatten();
+    (accounts.chain(token_accounts), last_notified_slot)
```

Comment:
This one is pretty hard to follow now. I guess the issue is with iterator types, but is there no better way to write this?

Reply:
Fair point. Switched to `Itertools::Either`. wdyt?

Reply:
Way better, thanks!